### DISCUSSION TITLE
Renaming rust modules

### DISCUSSION THREAD

**Discussion by mydumpfire** [helix-editor/helix#8923] [Troubleshooting]
Performing a rename on `mod foobar` or `use foobar` seems to rename the associated files, but not the references to it.

So if you have something like:

```
.
├── foobar.rs
├── lib.rs
└── main.rs
```

`lib.rs`:

```rust
pub mod foobar;
```

`main.rs`:

```rust
use my_lib::foobar;
```

Renaming the symbol `foobar` to `baz` will cause a file change:

```
.
├── baz.rs
├── lib.rs
└── main.rs
```

But all the references in the code stay the same:

`lib.rs`:

```rust
pub mod foobar;
```

I've tested this on VSCode (using the same toolchain) and it works as expected. But I don't think this is an LSP extension AFAIK. Interestingly, using the new `:move` command renames everything correctly.

**Comment by the-mikedavis** [comment]
This should be reported against rust-analyzer instead - rust-analyzer should be sending changes to that update the references.

**Reply by pascalkuthe** [reply]
The differences to vscode are likely caused by version mismatch. I even recently submitted related bugfixes upstream so if vscode is using the latest nightly and helix is using stable that would likely explain it

**Reply by mydumpfire** [reply]
Like I said in the post, I'm using the same toolchain that I used for Helix in VSCode. I've also tested this on neovim and renaming works correctly. I'm not sure what helix is doing differently here.

**Reply by mydumpfire** [reply]
To make it clearer, this is my setup:
helix: [b7f98d1](https://github.com/helix-editor/helix/commit/b7f98d1d9930b15f07342f67d2560aecd4ebc401)
rust: 1.76.0-nightly (9b13310ca 2023-11-24)

I'd appreciate it if you can at least reproduce this bug.

**Reply by archseer** [reply]
But what's your rust analyzer version? vscode installs and manages it's own install usually.

**Reply by mydumpfire** [reply]
I'm using the `rust-analyzer` that comes with the toolchain:
`rust-analyzer 1.76.0-nightly (6cf0888 2023-11-26)`
I've configured VSCode to use the same set of toolchain (verified with `ps`). Neovim also works correctly, btw (with the same toolchain of course).

**Comment by the-mikedavis** [comment]
Ah ok, I can reproduce this on rust-analyzer's current master (`c7c582afb57bb802715262d7f1ba73b8a86c1c5a`).

When we send the rename, rust-analyzer gives us only a resource operation to rename the file.

```
2023-11-28T18:26:24.400 helix_lsp::transport [INFO] rust-analyzer -> {"jsonrpc":"2.0","method":"textDocument/rename","params":{"newName":"baz","position":{"character":11,"line":0},"textDocument":{"uri":"file:///home/michael/tmp/src/main.rs"}},"id":2}
2023-11-28T18:26:24.401 helix_lsp::transport [INFO] rust-analyzer <- {"jsonrpc":"2.0","id":2,"result":{"documentChanges":[{"kind":"rename","oldUri":"file:///home/michael/tmp/src/foobar.rs","newUri":"file:///home/michael/tmp/src/baz.rs"}]}}
```

While Neovim gets a whole set of edits including the file rename operation:

```
[DEBUG][2023-11-28 18:47:31] .../vim/lsp/rpc.lua:284	"rpc.send"	{  id = 7,  jsonrpc = "2.0",  method = "textDocument/rename",  params = {    newName = "baz",    position = {      character = 11,      line = 0    },    textDocument = {      uri = "file:///home/michael/tmp/src/main.rs"    }  }}
[DEBUG][2023-11-28 18:47:31] .../vim/lsp/rpc.lua:387	"rpc.receive"	{  id = 7,  jsonrpc = "2.0",  result = {    documentChanges = { {        edits = { {            newText = "baz",            range = {              ["end"] = {                character = 15,                line = 0              },              start = {                character = 9,                line = 0              }            }          } },        textDocument = {          uri = "file:///home/michael/tmp/src/main.rs",          version = 0        }      }, {        edits = { {            newText = "baz",            range = {              ["end"] = {                character = 14,                line = 0              },              start = {                character = 8,                line = 0              }            }          } },        textDocument = {          uri = "file:///home/michael/tmp/src/lib.rs"        }      }, {        kind = "rename",        newUri = "file:///home/michael/tmp/src/baz.rs",        oldUri = "file:///home/michael/tmp/src/foobar.rs"      } }  }}

```

Neovim doesn't declare willRename support in its capabilities so it gets a different set of edits from rust-analyzer. If you comment this line: https://github.com/helix-editor/helix/blob/b023faacf898e2f393fa8f7f91e685547080442c/helix-lsp/src/client.rs#L562

then Helix and nvim behave the same.

Since we declare willRename support, we need to use it to handle rename resource operations from text document edits. I'll add an issue with the details. (https://github.com/helix-editor/helix/issues/8942)
